design/helm.md

```diff
@@ -0,0 +1,145 @@
+
+
+## Goals
+- Allow any Helm chart to be specified as part of a Kurl spec.
+
+## Non Goals
+
+- Verify authenticity, trustworthyness and "goodness" of helm addons.
+- Validate Helm values.
+
+## Background
+
+Users of Kurl have a limited set of available cluster addons and often want to include ones that are unsupported.
+There is friction in the [kurl development process](https://github.com/replicatedhq/kURL/blob/master/test/dev.md), [contributing new addons](https://kurl.sh/docs/add-on-author/) and addon maintanence. 
+
+Helm is a popular tool for managing deployment configuration that allows application/addon developers or the community to maintain a consistent interface for deployment through a chart.
+By allowing Kurl users to specifify a chart and its values, we can allow customization of kutl with new addons while delegating addons support and maintance to its community.
+
+## High-Level Design
+
+Kurl will bundle the Helm binary into the common tools bundle already delivering krew and kubectl plugins. The install.sh script will perform the following in the online case:
+1. Unarchive Helm and move into the `$PATH` and `/usr/bin`
+1. For each release, check if the release exists. 
+1. If the release exists, perform a `helm upgrade`.
+1. If the release doesn't exist, perform a `helm install`
+
+For airgapped the chart and artifacts need to be packaged into a bundle. The builder should perform the following:
+1. `helm pull` the repo
+1. `helm template` to render the chart with the provided values
+1. Generate an adhoc `manifest` file by parsing `image` specs from the rendered yaml.
+1. Generate a bundle including the chart + containers
+
+In the airgapped case, install.sh will perform the following:
+1. Unpack the helm bundle on initialization
+1. Load docker containers as part of the initialization
+1. Unarchive the chart and install with the installer spec values during the install phase
+
+## Detailed Design
+
+The yaml design is informed by the [Flux Helm Operator CRD](https://github.com/fluxcd/helm-operator/blob/master/docs/references/helmrelease-custom-resource.md). 
+Another spec is [Rancher's Helm CRD](/var/lib/rancher/rke2/server/manifests/) that works with k3s and RKE2.
+
+The proposal is to implement option1 in the updated Kurl Installer Spec:
+```yaml
+apiVersion: "cluster.kurl.sh/v1beta1"
+kind: "Installer"
+metadata: 
+  name: "latest"
+spec: 
+  kubernetes: 
+    version: "latest"
+  weave: 
+    version: "latest"
+  containerd: 
+    version: "latest"
+  helm:                 # Optional Future Capability
+    version: "latest"
+  helmReleases:
+    # Basic Idea, Option 1
+    - release: nginx-ingress
+      namespace: kube-system    # Optional
+      chart: 
+        name: bitnami/nginx
+        repo: https://charts.bitnami.com/bitnami 
+        version: 8.5.1          # Optional - uses latest if not provided
+      values: |                 # Optional
```

Comment:
it would be awesome if you somehow didnt have to include this in your spec. i cant really think of how. possibly a url but that would not work for airgap

Reply:
I think this one was resolved below. I think we still need to support this option so we don't need to answer "what url" and "where do I put my file".

Reply:
Why wouldn't you want to include this in your spec? Isn't that kind of the whole point? Are you saying that we would want to have some opinionated defaults? The point is to leverage the opinions of the underlying addon provider as much as possible, thus leveraging their helm chart defaults.

Reply:
Or are you just saying it should be specified in some other file? I like the simplicity and clarity of having a single file.

Reply:
my thought was allowing for opinionated defaults

Reply:
I was thinking of just having an external file here since the values can get crazy or we maybe there is a case where the installer is the same but the values are different per install. K3OS does the same thing with `k3os.install.config_url`

Reply:
So interesting, but perhaps decoupleable case. As in, replacing our opinionated addons with versions installed via Helm makes sense. But I see that as different from the current experimentation which is about giving vendors control. It does make sense to solve both with the same tooling though, so I see where the question comes from.